### PR TITLE
pfe-cta bug fixes + open brand styles

### PR DESCRIPTION
**Pull request by castastrophe**
https://rally1.rallydev.com/#/detail/userstory/305092942296?fdp=true

- https://github.com/patternfly/patternfly-elements/issues/294
- https://github.com/patternfly/patternfly-elements/issues/298
- https://github.com/patternfly/patternfly-elements/issues/236

### CODE REVIEW

**Comment by starryeyez024** [comment]
apologies in advance b/c i dont have time to checkout this branch & look, but wanted to capture somewhere that the current CTA component in the master branch has CSS such that the arrow breaks to a new line by itself. Not sure if you already solved for that or not #forgiveme 
image

**Comment by castastrophe** [comment]
@starryeyez024 I'm confused - the CTA buttons don't have the right arrow on them...

File: package.json [hunk 1]
```diff
@@ -81,7 +81,7 @@
     "gulp-trim": "^1.0.2",
     "gulp-uglify": "^3.0.1",
     "husky": "^1.1.1",
-    "lerna": "^3.0.0-rc.0",
+    "lerna": "^3.14.1",
```
Comment (mwcz) [inline package.json]: Upgrading lerna and browser-sync sounds good, but can we pull these into a separate PR?

File: elements/pfe-cta/src/pfe-cta.scss [hunk 2]
```diff
@@ -1,200 +1,344 @@
...
+  background-color: #{pfe-local(background)};
+
+  padding:       #{pfe-local(Padding)};
+  border-radius: #{pfe-local(BorderRadius)};
+  border:        #{pfe-var(ui--border-width)} #{pfe-var(ui--border-style)} #{pfe-local(border)};
+
+  // Update the broadcasted variables, map to main color
+  --pfe-broadcasted--color--ui-link: #{pfe-local(Color)};
+  --pfe-broadcasted--color--ui-link--hover: #{pfe-local(Color--hover)};
+  --pfe-broadcasted--color--ui-link--focus: #{pfe-local(Color--focus)};
```
Comment (starryeyez024) [inline elements/pfe-cta/src/pfe-cta.scss]: Yay!!! this works great. :)

Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.scss]: Ok fleeting friday afternoon thought: this does not solve for this scenario:
```
    <pfe-band pfe-color="darkest"> <-- should pass through the lighter blue broadcasted var -->
        <pfe-cta ><a href="#">Default</a></pfe-cta> <--overrides it to basic link blue -->
    </pfe-cta>
```

image

File: elements/pfe-cta/src/pfe-cta.js [hunk 3]
```diff
@@ -13,42 +13,94 @@ class PfeCta extends PFElement {
...
     const firstChild = this.children[0];
 
-    if (!firstChild) {
-      console.warn(
-        `${
-          PfeCta.tag
-        }:The first child in the light DOM must be an anchor tag (<a>)`
-      );
-    } else if (firstChild && firstChild.tagName.toLowerCase() !== "a") {
+    if (!firstChild || (firstChild && !["A", "BUTTON", "INPUT"].includes(firstChild.tagName))) {
```
Comment (starryeyez024) [inline elements/pfe-cta/src/pfe-cta.js]: @castastrophe this is throwing errors in IE11; when I revert it back to the original code it works. Can we back this out & make it a todo for later?

File: elements/pfe-cta/src/pfe-cta.js [hunk 4]
```diff
@@ -13,42 +13,94 @@ class PfeCta extends PFElement {
...
+  get defaultStyle() {
+    return this.hasAttribute("priority") ? false : true;
+  }
+
+  static get observedAttributes() {
+    return [
+      "pfe-priority",
+      "pfe-color",
+      "pfe-alt",
+      "on"
```
Comment (starryeyez024) [inline elements/pfe-cta/src/pfe-cta.js]: we need to remove alt & on @chrisdo1

File: elements/pfe-sass/README.md [hunk 5]
```diff
@@ -1,3 +1,68 @@
...
+4. Reset local variable values as needed for color attribute overrides.
+
+
+### CSS Example
+
+
+```
+    // 1. set up local vars equal to theme vars & fallbacks
+    :host {
+      --pfe-local--Color: var(theme--ui-link, #06c);
```
Comment (starryeyez024) [inline elements/pfe-sass/README.md]: make example clearer that pfe-local means somethign like pfe-cta

Comment (starryeyez024) [reply elements/pfe-sass/README.md]: for example if you are in CTA... its pfe-cta  @starryeyez024

File: elements/pfe-cta/src/pfe-cta.json [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            "primary",
+            "secondary"
+          ],
+          "observer": "_basicAttributeChanged"
+        },
+        "color": {
+          "title": "Color",
+          "type": "string",
+          "enum": [
+            "base",
```
Comment (starryeyez024) [inline elements/pfe-cta/src/pfe-cta.json]: need to update these; remove base, add lightest

File: elements/pfe-cta/src/pfe-cta.scss [hunk 7]
```diff
@@ -1,200 +1,367 @@
...
-  //  &::after {
-  //    @include pfe-arrow-color(var(--pfe-cta--main--visited));
-  //  }
-  // }
+  position: relative;
+  transition: all #{pfe-var(animation-timing)};
 
-  ::slotted(a:hover) {
-    color: var(--pfe-cta--main--hover);
+  line-height: inherit;
```
Comment (starryeyez024) [inline elements/pfe-cta/src/pfe-cta.scss]: on line 39, we should update `all` and change it to `padding`

File: elements/pfe-cta/src/pfe-cta.js (hunk 4 above)
Comment (starryeyez024) [inline elements/pfe-cta/src/pfe-cta.js]: list should be
```
      "priority",
      "color",
```

Comment (chrisdo1) [reply elements/pfe-cta/src/pfe-cta.js]: @starryeyez024 I actually had removed both of those but had not committed yet. Changed `pfe-color` to `color` as well.

Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.js]: or did we decide to keep `pfe-` namespacing to avoid conflicts with angular , etc?
@castastrophe @chrisdo1 @markcaron @kylebuch8

Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.js]: this issue says keep namespacing https://github.com/patternfly/patternfly-elements/issues/220

Comment (castastrophe) [reply elements/pfe-cta/src/pfe-cta.js]: I'm so torn.  Removing the namspacing is cleaner for developers.  Angular namespaces their stuff with `ng-` though.  I defer to others.  I like it without but can see the value of keeping the namespacing too.

Comment (chrisdo1) [reply elements/pfe-cta/src/pfe-cta.js]: Do we need to change the attributes in the light DOM then? 
`<pfe-cta priority="primary"><a href="#">Primary</a></pfe-cta>`
Should be:
`<pfe-cta pfe-priority="primary"><a href="#">Primary</a></pfe-cta>`

File: elements/pfe-cta/src/pfe-cta.js [hunk 8]
```diff
@@ -13,42 +13,94 @@ class PfeCta extends PFElement {
     return "pfe-cta.html";
   }
```
Comment (kylebuch8) [inline elements/pfe-cta/src/pfe-cta.js]: Did we intentionally leave `get schemaUrl() { return "pfe-cta.json"; }` out? If so, we should remove all of the `observedAttributes` and `attributeChangedCallback` code.

File: elements/pfe-cta/src/pfe-cta.js (hunk 3 above)
Comment (castastrophe) [reply elements/pfe-cta/src/pfe-cta.js]: Actually I can't replicate any errors.  I just pulled up the demo page in IE11 on BrowserStack and see no error messages in the console.

Comment (castastrophe) [reply elements/pfe-cta/src/pfe-cta.js]: I added a polyfill for includes because it's not supported in IE11 - let me know if that does not resolve the issue you are seeing.

File: elements/pfe-cta/src/pfe-cta.js (hunk 8 above)
Comment (castastrophe) [reply elements/pfe-cta/src/pfe-cta.js]: Thanks Kyle!  I added it with the latest commit.

File: elements/pfe-cta/src/pfe-cta.scss (hunk 7 above)
Comment (castastrophe) [reply elements/pfe-cta/src/pfe-cta.scss]: If we put the padding on the :host element, it works in IE11 but the default padding looks a little bigger.  Do you know why?

File: elements/pfe-cta/src/pfe-cta.js (hunk 4 above)
Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.js]: Since we talked about it as a larger group & it's [currently in the PFE documentation](https://github.com/patternfly/patternfly-elements-documentation/blob/master/content/develop/_index.md#attribute-naming), I think we should follow the prefix recommendation in both the JS file, the light DOM examples, everywhere (find & replace).

Comment (castastrophe) [reply elements/pfe-cta/src/pfe-cta.js]: I want to note that `on=` is the exception since that is being deprecated anyway

File: package.json (hunk 1 above)
Comment (starryeyez024) [reply package.json]: @mwcz sure i'll remove from this PR, not sure why its appearing here. Do you have this captured as an issue or something elsewhere?

File: elements/pfe-cta/src/pfe-cta.js (hunk 3 above)
Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.js]: same, no errors anymore, thx!

File: elements/pfe-cta/src/pfe-cta.scss (hunk 7 above)
Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.scss]: padding is on host now :)

File: elements/pfe-sass/variables/_colors.scss [hunk 9]
```diff
@@ -32,7 +32,7 @@ $pfe-color--ui-link--on-dark--visited:          lighten($pfe-color--ui-link--vis
 $pfe-color--ui-link--on-dark--hover:            lighten($pfe-color--ui-link--on-dark, 10%) !default; // Was #c0ecff
 $pfe-color--ui-link--on-dark--focus:            lighten($pfe-color--ui-link--on-dark, 10%) !default; // Was #c0ecff
 
-$pfe-color--ui-link--on-saturated:              lighten($pfe-color--ui-link, 40%) !default;               // Was #97daff
+$pfe-color--ui-link--on-saturated:              $pfe-color--white;               // Was #97daff
```
Comment (starryeyez024) [inline elements/pfe-sass/variables/_colors.scss]: This is a system change, but it makes more sense to me if no one objects?

**Comment by starryeyez024** [comment]
@castastrophe OK! I think this has been tidied, tested, browser tested, and is good to go!

File: elements/pfe-sass/variables/_colors.scss (hunk 9 above)
Comment (castastrophe) [reply elements/pfe-sass/variables/_colors.scss]: 👍

**Comment by starryeyez024** [comment]
I remember now why I had added the inner-border to the CTA secondary. The whole issue is surrounding the hover state on the secondary CTA when its on a dark background. What's complicated about this scenario is that the whole button is supposed to flip, where the background is white and the text is black. We need two colors then, white and black. Since we deprecated `on="dark"` , the only information the CTA has to know that it is on a dark background is the broadcast variables.  The problem is that the card only provides one real broadcast color which is white. So we don't actually have a variable available to use to set the text color with. 

We also have a mix'n'match approach to colors still, where some pfe-color attributes requesting surface colors and other attributes requesting ui colors:

image

**Comment by chrisdo1** [comment]
> I remember now why I had added the inner-border to the CTA secondary. The whole issue is surrounding the hover state on the secondary CTA when its on a dark background. What's complicated about this scenario is that the whole button is supposed to flip, where the background is white and the text is black. We need two colors then, white and black. Since we deprecated `on="dark"` , the only information the CTA has to know that it is on a dark background is the broadcast variables. The problem is that the card only provides one real broadcast color which is white. So we don't actually have a variable available to use to set the text color with.
> 
> We also have a mix'n'match approach to colors still, where some pfe-color attributes requesting surface colors and other attributes requesting ui colors, which might be unintuitive for people applying the attributes and guessing what color they are going to get.
> 
Thanks for reviewing, @starryeyez024. The first issue is how to address the secondary on dark backgrounds, which is the oddball of the cta world since the colors flip. Seems like the only solution would be to add new broadcast custom properties that would only affect secondary CTAs? It strikes me as a lot of overhead but what other choice do we have? The second issue is around the secondary cta color variables. We have most of our cta variables pointing to the `ui` color families. In the case of the secondary CTAs we are using the `surface` family to achieve the black color, since ui doesn't quite have anything close. Do we map the secondary color palette to the `ui-complement` family (and then change that in our theme repo)? `ui-complement` is quite a bit lighter than the specified color for secondary cta text and borders (#464646 compared to #151515).

File: elements/pfe-cta/src/pfe-cta.scss [hunk 10]
```diff
@@ -1,200 +1,359 @@
...
+  --pfe-cta--BackgroundColor:           transparent;
+  --pfe-cta--BorderColor:               var(--pfe-broadcasted--color--text, #{pfe-color(ui-complement)});
+  --pfe-cta__inner--BorderColor:        transparent;
+  --pfe-cta--Color:                     var(--pfe-broadcasted--color--text, #{pfe-color(ui-complement)});
+
+  --pfe-cta--BackgroundColor--hover:    #{pfe-color(ui-complement)};
+  --pfe-cta--BorderColor--hover:        var(--pfe-broadcasted--color--text, #{pfe-color(ui-complement)}); //white, then black
+  --pfe-cta--Color--hover:              #{pfe-color(ui-complement--text)};
+
+  --pfe-cta--BackgroundColor--focus:    #{pfe-color(surface--lighter)};
```
Comment (chrisdo1) [inline elements/pfe-cta/src/pfe-cta.scss]: @starryeyez024 this is the only color (it's a pale gray, #f0f0f0) for which we don't have anything close in the `ui-complement` family. Should I add a `ui-complement--focus` variable for this?

Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.scss]: I'm feeling like we could merge this as-is since it works, and then go back and update color variables here?

Comment (chrisdo1) [reply elements/pfe-cta/src/pfe-cta.scss]: Okay, works for me.

File: package.json (hunk 1 above)
Comment (mwcz) [reply package.json]: #90

File: examples/index.html [hunk 11]
```diff
@@ -22,7 +22,9 @@ <h1>Demos</h1>
 					<li><a href="../elements/pfe-icon/demo">pfe-icon</a></li>
 					<li><a href="../elements/pfe-icon-panel/demo">pfe-icon-panel</a></li>
 					<li><a href="../elements/pfe-layouts/demo">pfe-layouts</a></li>
+					<li><a href="../elements/pfe-link-list/demo">pfe-link-list</a></li>
 					<li><a href="../elements/pfe-markdown/demo">pfe-markdown</a></li>
+					<li><a href="../elements/pfe-navigation/demo">pfe-navigation</a></li>
```
Comment (mwcz) [inline examples/index.html]: The changes in this file look like they're holdovers from another branch.

File: elements/pfe-cta/src/pfe-cta.js [hunk 12]
```diff
@@ -13,42 +13,120 @@ class PfeCta extends PFElement {
...
+
+  _basicAttributeChanged(attr, oldValue, newValue) {
+    this[attr].value = newValue;
+  }
 }
 
 PFElement.create(PfeCta);
 
-export default PfeCta;
+export { PfeCta as default };
```
Comment (mwcz) [inline elements/pfe-cta/src/pfe-cta.js]: Just curious, why this change?

File: examples/index.html (hunk 11 above)
Comment (chrisdo1) [reply examples/index.html]: @mwcz Thanks for reviewing. Removed pfe-navigation and pfe-link-list from the examples index file.

File: elements/pfe-cta/src/pfe-cta.js (hunk 12 above)
Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.js]: not sure; @kylebuch8 ?

Comment (kylebuch8) [reply elements/pfe-cta/src/pfe-cta.js]: @KamiQuasi is requesting that we do this. And I've seen some posts on why this is better.

Comment (kylebuch8) [reply elements/pfe-cta/src/pfe-cta.js]: That said, should we just leave this change out of this PR and do one big change where we update them all at the same time?

Comment (starryeyez024) [reply elements/pfe-cta/src/pfe-cta.js]: removed

Comment (mwcz) [reply elements/pfe-cta/src/pfe-cta.js]: Cool.  I'm not that familiar with arguments for the `export { Foo as default }` style.  I am ready to be swayed.  I just like the sweet simple syntax of `export default Foo`.
